### ISSUE TITLE
dmrgscf freezes after start (too large integral ?)

### ISSUE THREAD

**Issue by leochaussy** [block-hczhai/block2-preview#28]
Hello, 
I come to you again with a question. I am trying to perform state average dmrgscf with pyscf (2.1.1) and the last release of block2 (not the mpi one) installed through pip on a rather big 1st row transition metal complex (triplet). I couldn't diagnose if I was hitting the limit of what is feasible with the current implementation or if I'm doing something wrong.
I have a "moderately" large active space (18e,20o) with a large basis set 745 cGTO (and no density fitting) and 30 atoms and a max starting bond dimension of 200 to try and get decent orbitals.
So far the program seems to freeze when calling block2main after the ROHF step.

Last test was on 16 threads and an exceedingly large amount of memory (1Tb), as issues related to this was my first guess. First it's using all the threads for ERI calculations and throughout the ROHF part. When entering the DMRG part only one cpu remains active and the ~230Gb used memory remains in use (i'm guessing integrals are stored in memory). After that nothing happens and it remains in that state (longest i've let it is about 12hours, i ended up killing it). I couldn't diagnose what is happening at that step since it seems to be at the limit between the two programs. I'm not sure what is happening there, i would expect a long integral transformation step but the fact that it runs only one core makes me doubt what's really going on.

And the Slurm scheduler seems to be lost too:
`slurmstepd: error: slurm_get_node_energy: Socket timed out on send/recv operation
slurmstepd: error: _get_joules_task: can't get info from slurmd
`
This is where it stops:
`******** <class 'pyscf.mcscf.addons._state_average_mcscf_solver.<locals>.StateAverageMCSCF'> ********
CAS (10e+8e, 20o), ncore = 54, nvir = 671
frozen orbitals 17
max_cycle_macro = 50
max_cycle_micro = 4
conv_tol = 1e-07
conv_tol_grad = None
orbital rotation max_stepsize = 0.02
orbital rotation threshold for CI restart = 0.01
augmented hessian ah_max_cycle = 30
augmented hessian ah_conv_tol = 1e-12
augmented hessian ah_linear dependence = 1e-14
augmented hessian ah_level shift = 1e-08
augmented hessian ah_start_tol = 2.5
augmented hessian ah_start_cycle = 3
augmented hessian ah_grad_trust_region = 3
kf_trust_region = 3
kf_interval = 4
ci_response_space = 4
ci_grad_trust_region = 3
with_dep4 0
natorb = True
canonicalization = True
sorting_mo_energy = False
ao2mo_level = 2
chkfile = model_tpa.chk
max_memory 900000 MB (current use 309642 MB)
internal_rotation = False

******** Block flags ********
executable             = /home/lchaussy/.local/bin/block2main
BLOCKEXE_COMPRESS_NEVPT= /home/lchaussy/.local/bin/block2main
`
[input.txt](https://github.com/block-hczhai/block2-preview/files/10326034/input.txt)

I join the minimal working example input that I used for debugging purposes.

I have run successfully the same script with minimal basis set, and also tested on smaller molecules with that same basis and input (with outstanding efficiency I have to say). Is this the limit of what is actually feasible or is my strategy wrong here? I am very new to the pyscf ecosystem and block and might be missing some obvious things. 
I did also see several strategies to reduce memory use in block, which I have not use here yet thinking that if a minimal bond dimension as this one (200) doesn't work out no need to push harder anyway.

**Comment by hczhai** [comment]
Thanks for your detailed description of the problem. As I personally never run calculations at this scale, let's figure out the problem together.

First we need some more information to identity whether the problem is related to ``block2`` or not.

(1) It should always be a good strategy to run CASCI before CASSCF to see if you can finish one iteration of CASSCF. You can simply change the "CASSCF" in your script to "CASCI".
(1.1) If this CASCI step has some problems, then it is likely the problem is related to ``block2``. If this is the case, you can follow the steps in https://block2.readthedocs.io/en/latest/user/basic.html#integral-generation to first use ``dryrun`` method to generate a FCIDUMP file, then run ``block2main`` manually. There, from the output of ``block2main`` we can get more information.
(1.2) Otherwise, if CASCI can run but CASSCF cannot, then it is likely the problem is that ``pyscf.mcscf`` has some problem for handling such a large basis. But sometimes a large basis with a smaller active space may work.

(2) If you run/test several jobs simultaneously, it is important to make sure that each job has a distinct scratch folder, namely, ``lib.param.TMPDIR``. Otherwise, they will influence each other and you will get unpredictable results.

**Comment by leochaussy** [comment]
Thank you for your fast response. I do have an independent scratch, all good on that part.

(1) I launched a CASCI which leads to same result, it stops here.

```(...)
SCF energy = -2257.60971995501

******** CASCI flags ********
CAS (10e+8e, 20o), ncore = 54, nvir = 671
natorb = True
canonicalization = True
sorting_mo_energy = False
max_memory 1000000 (MB)

******** Block flags ********
executable             = /home/lchaussy/.local/bin/block2main
BLOCKEXE_COMPRESS_NEVPT= /home/lchaussy/.local/bin/block2main
```

The fact that it gets stuck on one cpu at that spot makes me think of GIL, but I didn't really manage to get more out of it, even by hitting the max verbosity. Last obvious call seems to be related to the dump_flags() function in dmrgscf but from here I couldn't get much more.

(2) Trying the dry run test also leads to the same results (freeze at that exact point), I didn't push through however but Slurm kept going with more of :
`
slurmstepd: error: slurm_get_node_energy: Socket timed out on send/recv operation slurmstepd: error: _get_joules_task: can't get info from slurmd `

If this is purely pyscf or dmrgscf related I might as well restart the thread at the appropriate location. Otherwise may I use any FCIDUMP file to try and push block2 to the limit anyway ? I was thinking maybe using Openmolcas, Molpro...  does it need some additional formatting ?

**Comment by hczhai** [comment]
Thanks for the feedback.

It seems that ``CASSCF/CASCI`` has an independent ``verbose`` parameter. You may change ``mc.kernel()`` (``mc`` can be a CASSCF or CASCI object) to ``mc.kernel(verbose=5)`` and hopefully it will print more information after the ``dump_flags`` function. Then you can do the same thing with some smaller systems, and you can compare the output to see where the stuck location is.

``dryrun`` may not be helpful at this point since I do not know how to control its verbosity.

**Comment by leochaussy** [comment]
I retried a CASCI using `mc.kernel(verbose=5)` without success for the good reason that in seems to never enter anything which can print anythung. I added print statements here and there to figure out what happens last. It successfully enters in CASCI.kernel(), where it calls its own version of dump_flags. This should be related to the successful print of  ***CASCI flags *** as in my previous message. From there it calls dmrgci.dump_flags() and execute the two first statements but never goes beyond:
`        log.info('BLOCKEXE_COMPRESS_NEVPT= %s', settings.BLOCKEXE_COMPRESS_NEVPT)`
(I added print before and after every lines to be sure).

Interestingly, I also performed an additional try using density fitting. It manages to go one step beyond calls the block_function() and returns the following error:
```
Traceback (most recent call last):
  File "rohf.py", line 246, in <module>
    mc.kernel(verbose=5)
  File "/home/lchaussy/.local/lib/python3.8/site-packages/pyscf/mcscf/casci.py", line 946, in kernel
    self.dump_flags(log)
  File "/home/lchaussy/.local/lib/python3.8/site-packages/pyscf/mcscf/df.py", line 88, in dump_flags
    casscf_class.dump_flags(self, verbose)
  File "/home/lchaussy/.local/lib/python3.8/site-packages/pyscf/mcscf/casci.py", line 817, in dump_flags
    self.fcisolver.dump_flags(log.verbose)
  File "/home/lchaussy/.local/lib/python3.8/site-packages/pyscf/mcscf/addons.py", line 879, in dump_flags
    fcibase_class.dump_flags(self, verbose)
  File "/home/lchaussy/.local/lib/python3.8/site-packages/pyscf/dmrgscf/dmrgci.py", line 248, in dump_flags
    log.info('Block version          = %s', block_version(self.executable))
  File "/home/lchaussy/.local/lib/python3.8/site-packages/pyscf/dmrgscf/dmrgci.py", line 1022, in block_version
    msg = check_output([blockexe, '-v'], stderr=STDOUT).decode()
  File "/trinity/shared/apps/tr17.10/x86_64/python3-3.8.6/lib/python3.8/subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/trinity/shared/apps/tr17.10/x86_64/python3-3.8.6/lib/python3.8/subprocess.py", line 489, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/trinity/shared/apps/tr17.10/x86_64/python3-3.8.6/lib/python3.8/subprocess.py", line 854, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/trinity/shared/apps/tr17.10/x86_64/python3-3.8.6/lib/python3.8/subprocess.py", line 1637, in _execute_child
    self.pid = _posixsubprocess.fork_exec(
OSError: [Errno 513] Unknown error 513

```
Not sure where to go from there, I'm looking into the subprocess module but couldn't get much from it. There doesn't seem to be an issue with the "block2main -v" command executed in check_output() (I can reproduce the steps and execute it in a shell on the same machine).

**Comment by hczhai** [comment]
Maybe you can try adding ``dmrgscf.settings.BLOCKVERSION = '2.0'`` at the beginning of your script so that it can skip the ``check_output`` part?

**Comment by hczhai** [comment]
Note that when you run your python script, there should not be any ``srun`` or ``mpirun`` before the ``python`` command. If unclear, attaching your job submission script might be helpful.

**Comment by leochaussy** [comment]
Setting `dmrgscf.settings.BLOCKVERSION = '2.0'` allowed to bypass the problem. In the density fitting case it leads to a similar error following the writeIntegralFile() call. 
 It's going a little further now but gets stopped again due to another call to subprocess. 
If I don't use density fitting it goes up til the same point with all 16 threads (effective h1e calculation) and then seems to hang there on one thread same as before (without error though).
 
 Here is the output and the error for the Density-Fitting case. 

```
******** CASCI flags ********
CAS (10e+8e, 20o), ncore = 54, nvir = 671
natorb = True
canonicalization = True
sorting_mo_energy = False
max_memory 1000000 (MB)

******** Block flags ********
executable             = /home/lchaussy/.local/bin/block2main
BLOCKEXE_COMPRESS_NEVPT= /home/lchaussy/.local/bin/block2main
Block version          = 2.0
mpiprefix              = 
scratchDirectory       = /scratch/lchaussy/pyscf/model_tpa/testjob/git_debug/verbose_restart/print_restart/dfit/restartblock2version/tmp
integralFile           = /scratch/lchaussy/pyscf/model_tpa/testjob/git_debug/verbose_restart/print_restart/dfit/restartblock2version/tmp/FCIDUMP
configFile             = /scratch/lchaussy/pyscf/model_tpa/testjob/git_debug/verbose_restart/print_restart/dfit/restartblock2version/tmp/dmrg.conf
outputFile             = /scratch/lchaussy/pyscf/model_tpa/testjob/git_debug/verbose_restart/print_restart/dfit/restartblock2version/tmp/dmrg.out
maxIter                = 26
scheduleSweeps         = [0, 2, 4, 6, 8, 10, 12, 14]
scheduleMaxMs          = [200, 200, 200, 200, 200, 200, 200, 200]
scheduleTols           = [0.0001, 1e-05, 1.0000000000000002e-06, 1.0000000000000002e-07, 1.0000000000000002e-08, 1.0000000000000003e-09, 1.0000000000000003e-10, 1.0000000000000001e-11]
scheduleNoises         = [0.0001, 1e-05, 1.0000000000000002e-06, 1.0000000000000002e-07, 1.0000000000000002e-08, 1.0000000000000003e-09, 1.0000000000000003e-10, 0.0]
twodot_to_onedot       = 18
tol                    = 1e-10
maxM                   = 200
dmrg switch tol        = 0.001
wfnsym                 = 1
fullrestart            = False
num_thrds              = 16
memory                 = 1000

State-average over 11 states with weights [0.09090909 0.09090909 0.09090909 0.09090909 0.09090909 0.09090909
 0.09090909 0.09090909 0.09090909 0.09090909 0.09090909]
DFCASCI/DFCASSCF: density fitting for JK matrix and 2e integral transformation
Start CASCI
    CPU time for integral transformation to CAS space     14.58 sec, wall time      0.91 sec
    CPU time for jk     20.23 sec, wall time      1.46 sec
    CPU time for jk     19.29 sec, wall time      1.39 sec
    CPU time for jk     19.40 sec, wall time      1.40 sec
    CPU time for jk     19.44 sec, wall time      1.42 sec
    CPU time for jk     19.53 sec, wall time      1.52 sec
    CPU time for jk     19.50 sec, wall time      1.51 sec
    CPU time for jk     19.46 sec, wall time      1.49 sec
    CPU time for jk     19.43 sec, wall time      1.49 sec
    CPU time for jk     19.48 sec, wall time      1.50 sec
    CPU time for jk     19.44 sec, wall time      1.50 sec
    CPU time for jk     19.44 sec, wall time      1.50 sec
    CPU time for jk     19.44 sec, wall time      1.49 sec
    CPU time for jk     19.44 sec, wall time      1.48 sec
    CPU time for jk     19.41 sec, wall time      1.49 sec
    CPU time for jk     19.40 sec, wall time      1.49 sec
    CPU time for jk     19.40 sec, wall time      1.50 sec
    CPU time for jk      4.83 sec, wall time      0.39 sec
    CPU time for df vj and vk    316.60 sec, wall time     24.02 sec
core energy = -2217.27626887629
    CPU time for effective h1e in CAS space    327.92 sec, wall time     32.93 sec

```

And the following error:

```
Traceback (most recent call last):
  File "rohf.py", line 247, in <module>
    mc.kernel(verbose=5)
  File "/home/lchaussy/.local/lib/python3.8/site-packages/pyscf/mcscf/casci.py", line 949, in kernel
    kernel(self, mo_coeff, ci0=ci0, verbose=log)
  File "/home/lchaussy/.local/lib/python3.8/site-packages/pyscf/mcscf/casci.py", line 577, in kernel
    e_tot, fcivec = casci.fcisolver.kernel(h1eff, eri_cas, ncas, nelecas,
  File "/home/lchaussy/.local/lib/python3.8/site-packages/pyscf/mcscf/addons.py", line 891, in kernel
    e, c = fcibase_class.kernel(self, h1, h2, norb, nelec, ci0=ci0,
  File "/home/lchaussy/.local/lib/python3.8/site-packages/pyscf/dmrgscf/dmrgci.py", line 724, in kernel
    writeIntegralFile(self, h1e, eri, norb, nelec, ecore)
  File "/home/lchaussy/.local/lib/python3.8/site-packages/pyscf/dmrgscf/dmrgci.py", line 944, in writeIntegralFile
    check_call(cmd, shell=True)
  File "/trinity/shared/apps/tr17.10/x86_64/python3-3.8.6/lib/python3.8/subprocess.py", line 359, in check_call
    retcode = call(*popenargs, **kwargs)
  File "/trinity/shared/apps/tr17.10/x86_64/python3-3.8.6/lib/python3.8/subprocess.py", line 340, in call
    with Popen(*popenargs, **kwargs) as p:
  File "/trinity/shared/apps/tr17.10/x86_64/python3-3.8.6/lib/python3.8/subprocess.py", line 854, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/trinity/shared/apps/tr17.10/x86_64/python3-3.8.6/lib/python3.8/subprocess.py", line 1637, in _execute_child
    self.pid = _posixsubprocess.fork_exec(
OSError: [Errno 513] Unknown error 513
```

I also attach the submit script in case but i'm using sbatch and nothing before the python call (with sbatch script.sh).
[submit.txt](https://github.com/block-hczhai/block2-preview/files/10327438/submit.txt)

**Comment by leochaussy** [comment]
I managed to do a temporary fix of the problem by removing the two following lines in dmrgci.py
```
    cmd = ' '.join((DMRGCI.mpiprefix, "mkdir -p", DMRGCI.scratchDirectory))
    check_call(cmd, shell=True)
```
This allows generation of FCIDUMP. The script crashes again shortly after on calling once more check_call() to execute Block but I can at least launch Block2 directly without issue (at least in the density fitting case, I will do more test without). All seems to run fine on that side it starts the sweep directly.

**Comment by leochaussy** [comment]
Rebuilding a clean conda environment using Python 3.9.15 totally fixed the problem... I was using regular python 3.8.6 for information... (although it might not be just that, maybe something to do with our system also)  I see some users experienced issues with subprocess in the past on the pyscf repository but I don't have a good explanation for all that. Using the shell=True statement in subprocess functions like check_call() or check_output() can be a little dangerous for other reasons mentioned here and there since it loads a shell and allows to feed all sort of possibly unwanted stuff but I don't think that was really the problem for me, I checked my bash profile and cie and nothing seems related to that. Anyway thanks a lot for the help. It is up and running now .

**Comment by hczhai** [comment]
Thanks for the detailed update. It is great that the problem was solved. This is very informative for me and I believe it will also help many other users.

So the issue seems to be related to some specific design choices (using subprocesses) made in ``pyscf.dmrgscf``, which is written before the "block2" project started. I totally agree that subprocess is not a good strategy in today's software standards. In ``block2`` we tried very hard to decrease the level of software dependence and complexities, and in this block2 repo there is no usage of subprocesses. The file ``pyblock2/dmrgscf.py`` provided in ``block2``, that I mentioned earlier, is in fact an alternative DMRGSCF implementation without any usage of subprocesses. But some of its parameters are over-simplified there so it is not good for production runs. 

By the way, in the example script I provided in the previous issue I used 1E-10 as the tolerance to increase the reproducibility for small systems, but this may not be a good choice for production runs / large systems. The default tolerance may be more appropriate.

Good luck and happy new year!
